File: Natas/L11/L11_decode.php
#!/usr/bin/php

<?php

// ? 解xor_encrypt() key 想法
// ? => cheapText ^ key = ciphertext
// ? => cheapText ^ ciphertext = key

function xor_encrypt($in, $key)
{
    $text = $in;
    $outText = '';

    // Iterate through each character
    for ($i = 0; $i < strlen($text); $i++) {
        $outText .= $text[$i] ^ $key[$i % strlen($key)];
    }

    return $outText;
}

// TODO: Test json_encode()
$defaultdata = array("showpassword" => "no", "bgcolor" => "#ffffff");
// echo json_encode($defaultdata);

// TODO: Get key && test key
$data = base64_decode("ClVLIh4ASCsCBE8lAxMacFMZV2hdVVotEhhUJQNVAmhSEV4sFxFeaAw=");
$defaultdata = json_encode($defaultdata);
// ? substr($str, start, length);
$key = substr(xor_encrypt($data, $defaultdata), 0, 4);
// $cheapText = xor_encrypt($data, $key);
// echo $cheapText;
// echo "\n";
// echo $key;

// TODO: Encrypt cheapText to cipherText
$cheapText = json_encode(array("showpassword" => "yes", "bgcolor" => "#ffffff"));
$cipherText = base64_encode(xor_encrypt($cheapText, $key));
echo $cipherText;

File: Natas/L11/L11_decode.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import base64
import json

# ? cheapText ^ key = ciphertext
# ? => cheapText ^ ciphertext = key


def xor_encrypt(inp, key):
    text = inp
    outText = ''

    for i in range(len(text)):
        s = chr(ord(text[i]) ^ ord(key[i % len(key)]))
        outText += s
    return outText


def get_xor_encrypt_key(cipherText, defaultdata):
    key = ''
    repeat_key = xor_encrypt(cipherText, defaultdata)

    for i in repeat_key:
        if i in key:
            break
        key += i

    return key


def key_test(cipherText, defaultdata):
    key = get_xor_encrypt_key(cipherText, defaultdata)
    cheapText = xor_encrypt(cipherText, key)
    if cheapText == defaultdata:
        return True
    else:
        return False


# TODO: Get xor_encrypt() key
encrypt_cipher = base64.b64decode(
    "ClVLIh4ASCsCBE8lAxMacFMZV2hdVVotEhhUJQNVAmhSEV4sFxFeaAw=").decode('utf-8')
# ! 因為我的VSCode會在我存檔的時候自動排版，會多加了空格，所以用.replace()把空格去掉
# ? .dumps() => 將 Python 對象編碼成 JSON 字符串
defaultdata = json.dumps(
    {"showpassword": "no", "bgcolor": "#ffffff"}).replace(' ', '')

if (key_test):
    key = get_xor_encrypt_key(encrypt_cipher, defaultdata)
else:
    assert "The key is not currect!!!"


# TODO: Encrypt cheapText to ciphertext
cheapText = json.dumps(
    {"showpassword": "yes", "bgcolor": "#ffffff"}).replace(' ', '')
encrypt = xor_encrypt(cheapText, key).encode('utf-8')
cipherText = base64.b64encode(encrypt).decode('utf-8')
print(cipherText)
